### ISSUE TITLE
Crash: EXC_BAD_ACCESS attempting to evaluate a backtrace

### ISSUE THREAD

**Issue by bdkjones** [sass/libsass#2345]
### Title: The Sass provided at this link 100% consistently crashes Libsass 3.5.0 Beta 3. 

Example Sass: http://codepen.io/jakob-e/pen/doMoML

I'm compiling with the Libsass C API, but I do not believe my implementation is the cause of the crash because the backtrace shows that Libsass is finding "invalid sass" before the crash. The exception is a BAD_ACCESS on line 70 of `eval.cpp`, here:

<img width="1126" alt="screen shot 2017-03-01 at 16 43 55" src="https://cloud.githubusercontent.com/assets/926566/23488024/daba9414-fe9e-11e6-9fee-13ce20425b89.png">

And the backtrace for how it got there is:

<img width="868" alt="screen shot 2017-03-01 at 16 44 10" src="https://cloud.githubusercontent.com/assets/926566/23488030/e31d0dda-fe9e-11e6-9c6a-53a428ad684a.png">

I can provide more info if needed. Thanks!

version info:
```
libsass         3.5.0 Beta 3 (compiled with Clang on macOS 10.12.3)
```

**Comment by mgreter** [comment]
Pretty sure you are hitting stack size limitations (and therefore getting a [stack overflow exception, which we can't really catch][1]). This can happen on any function call, i.e. as with `exp.backtrace()` in your case. I guess the default on your platform is 1MB stack size (or even less). Your given [sample works (errors) correctly][2] (verified on windows) when the stack size is increased to 4MB (e.g. with MSVC `/STACK:"4194304"`). So please make sure to increase the stack size to a reasonable amount when compiling your implementation!

[1]: http://stackoverflow.com/questions/1578878/catching-stack-overflow-exceptions-in-recursive-c-functions
[2]: http://libsass.ocbnet.ch/srcmap/#QGltcG9ydCAiaHR0cDovL3d3dy5vY2JuZXQuY2gvYnVncmVwb3J0L2xpYnNhc3MvaXNzdWUtMjM0NS5zY3NzIjs=

**Comment by bdkjones** [comment]
Interesting. According to http://stackoverflow.com/questions/18909395/how-do-i-increase-the-stack-size-when-compiling-with-clang-on-os-x, the default stack size for Clang on macOS is `8MB`. I'll verify.

**Comment by bdkjones** [comment]
Ok, I verified that the process running libsass has an `8MB` stack. I then raised this to `64MB`, which is the maximum value permitted by macOS. I verified that the process had a `64MB` stack as shown below using `getrlimit()` and `setrlimit()`. 

I'm still getting the exception, though. I'm totally willing to admit this is related to the compiling options I'm using, but do you have any suggestions for what else to check?

**Comment by bdkjones** [comment]
Yea, the linker flag is `-Wl,-stack_size,1000000` for a `16MB` stack. But you can also raise it dynamically at runtime using `setrlimit()`. I've done both. I'll keep playing around, but the tool where you're verifying this is running Libsass 3.4.xx. Is there a way to test with that tool using the same build of Libsass that I'm using? (The 3.5.0 Beta 3 release).

**Comment by mgreter** [comment]
FWIW: I might be wrong here, but I would expect `setrlimit` to be a limit on top of the linked flag. I don't think (at least would not assume) that this works, so make sure you actually linked with that flag and make sure all other possible limits (ulimits, selinux etc) are big enough. Maybe you get something in a error.log?

I can see if I can update the version of the srcmap inspector ...

**Comment by bdkjones** [comment]
Yea, I'm positive the linker flag is getting through. Here's the result, showing the libsass-service has a stack size of `64MB`:

![screen shot 2017-03-01 at 22 29 31](https://cloud.githubusercontent.com/assets/926566/23495621/acaec790-fece-11e6-8165-030e9a8bf96e.png)

My current theory is that while the *main* thread of this process has the huge stack, the secondary thread where Libsass is actually running has been set up with a `512kb` stack. I'm using Apple's XPC framework to run Libsass in an isolated process and I'm not *super* familiar with the low-level details of how XPC manages its threads---you don't get to control how XPC spawns and runs threads; that's all implementation details. But that *would* explain this behavior so I'm looking into it.

I'll go ahead and close though, because this doesn't sound like a Libsass issue.

**Comment by bdkjones** [comment]
*sigh*. Confirmed. If I take XPC out of the equation and just run Libsass on the main thread of my app, this issue disappears. The child process in XPC is spawning custom threads to do its work and its setting the stack size of that thread to `512kb`. The `8MB` default applies only to main threads in macOS.

Dunno how to combat it yet. Have I told you guys how much I love programming recently?

**Comment by mgreter** [comment]
Sound very similar to https://github.com/libuv/libuv/issues/669, maybe that'll help! Good luck!

**Comment by bdkjones** [comment]
Yea, I saw that libuv bug while googling earlier. Their solution appears to be the only one available: create your own thread and set the stacksize manually before creation (it can't be changed once the thread is running). That's extra overhead because it has to go down into the kernel, but I can't find any other workaround. I have a SO question posted and I'll file a Radar with Apple --> http://stackoverflow.com/questions/42551114/xpc-service-raise-stack-size-on-worker-thread

**Comment by mgreter** [comment]
> Dunno how to combat it yet. Have I told you guys how much I love programming recently!

Not sure if it helps, but that is how you know that you are on a frontier, where no one else has been before. When I implement new things I always expect to find these kind of bugs in dependencies, as I did recently with [webgl][1] (took me around 3 days to figure out the real issue). When I started coding in 1994, browsers had bugs no one would ever expect or accept nowadays, and the only debugging we had was alert! None the less we were able to code around it and still produce nice results! So hang in there!

[1]: https://bugzilla.mozilla.org/show_bug.cgi?id=1343705
